x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/aggregations/metrics/GeoShapeBoundsAggregator.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.spatial.aggregations.metrics;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.search.aggregations.Aggregator;
+import org.elasticsearch.search.aggregations.LeafBucketCollector;
+import org.elasticsearch.search.aggregations.LeafBucketCollectorBase;
+import org.elasticsearch.search.aggregations.metrics.GeoBoundsAggregator;
+import org.elasticsearch.search.aggregations.support.ValuesSource;
+import org.elasticsearch.search.internal.SearchContext;
+import org.elasticsearch.xpack.spatial.index.mapper.GeoShapeValuesSource;
+import org.elasticsearch.xpack.spatial.index.mapper.MultiGeoShapeValues;
+
+import java.io.IOException;
+import java.util.Map;
+
+public final class GeoShapeBoundsAggregator extends GeoBoundsAggregator {
+    private final GeoShapeValuesSource valuesSource;
+
+    public GeoShapeBoundsAggregator(String name, SearchContext aggregationContext, Aggregator parent,
+                             GeoShapeValuesSource valuesSource, boolean wrapLongitude, Map<String, Object> metadata) throws IOException {
+        // the valuesSource as GeoPoint passed to GeoBoundsAggregator is not used
+        super(name, aggregationContext, parent, ValuesSource.GeoPoint.EMPTY, wrapLongitude, metadata);
```

Comment:
@iverase I really dislike this hack, but I am also weary of editing GeoBoundsAggregator to be more generic just for this extension. I might take another pass at this tomorrow

Reply:
Yes, it seems that `GeoBoundsAggregator` should be an abstract class and then having two implementation one for points and one for shapes?

Another thing I notice is that there is special logic when valueSource == null, cannot we have an empty bounds aggregation aggregation that is returned in that case when calling the register?

Reply:
I believe what will need to happen is that I make `ValuesSource.GEOPOINT` abstract as was done in  the previous branch. but it is not needed for the GeoBounds aggregator. may happen for the GeoGrid. to help resolve the CellIdSource work. For now, I prefer to have duplicate code and inherit from MetricsAggregator instead of GeoShapeAggregator and undo this hack

Reply:
>For now, I prefer to have duplicate code and inherit from MetricsAggregator

Having duplicate code seems to increase the likelihood these diverge, which would mean inconsistent behavior across oss and default distributions right? IMO in order to keep the behavior consistent, modeling all the default distribution overrides as pure subclasses of oss will minimize the possibility the implementations diverge.

Reply:
hmmm. OK. I have been a little resistant to introducing abstractions in server that only have one implementation in server. I'll overcome this and re-introduce the generic valuessource type. This should re-enable more proper inheritance here.

Reply:
after speaking with Ryan on videochat, we realized that we were misunderstanding one another. We are not duplicating the whole aggregation, but instead just the aggregator logic. We decided that this is acceptable and I'll continue with this approach. There will be some boilerplate that can be shared between, but that does not warrant any hackery